src/repl.ml

```diff
@@ -0,0 +1,191 @@
+open Import
+
+module ReplPath = struct
+  type t = string option
+
+  let of_json json =
+    let open Jsonoo.Decode in
+    nullable string json
+
+  let to_json (t : t) =
+    let open Jsonoo.Encode in
+    nullable string t
+
+  let key = "path"
+
+  let t = Settings.create ~scope:Global ~key ~of_json ~to_json
+end
+
+module ReplArgs = struct
+  type t = string list option
+
+  let of_json json =
+    let open Jsonoo.Decode in
+    nullable (list string) json
+
+  let to_json (t : t) =
+    let open Jsonoo.Encode in
+    nullable (list string) t
+
+  let key = "args"
+
+  let t = Settings.create ~scope:Global ~key ~of_json ~to_json
+end
+
+let get_repl_path () =
+  Option.join (Settings.get ~section:"ocaml.repl" ReplPath.t)
+
+let get_repl_args () =
+  Option.join (Settings.get ~section:"ocaml.repl" ReplArgs.t)
+
+let name = "REPL"
+
+let has_utop sandbox =
+  let open Promise.Syntax in
+  let cmd = Sandbox.get_command sandbox "utop" [ "-version" ] in
```

Comment:
This check isn't quite right for `$ dune utop` since we don't use the binary. In any case, I suggest that we use `Bin.which "utop"` to make a check like this (as we do elsewhere).

Reply:
`Cmd.check` is global so doesn't do exactly what we want here. Let's say the user has utop in its path (globally installed, or in the global switch), `Cmd.check` will return `Ok`, but `dune utop` and `opam exec -- utop` would fail.

Reply:
Yes, but you can fix this by passing the PATH of the sandbox to which.

Reply:
True, although we'd need to get the env differently for Esy, and we'd need to run a command (i.e. `opam env` or the esy equivalent) and parse them to access the env of the sandbox. The benefit of `Cmd.which` is that we can save a command execution, which won't be the case in this specific case.

Reply:
> although we'd need to get the env differently for Esy, and we'd need to run a command (i.e. opam env or the esy equivalent) and parse them to access the env of the sandbox

We had code that did this before. I wonder where it went 🤔 In any case it would be useful to have it.

> The benefit of Cmd.which is that we can save a command execution, which won't be the case in this specific case.

My goal isn't to shave off a command invocation, but to stop relying on the command lines of these tools and treat them all consistently. However, I just checked and we already rely on `--version` for lsp, so I suppose we can just leave this be and get rid of this stuff another time.

Reply:
> We had code that did this before. I wonder where it went

I did something to use a custom `env` for `Cmd.check` and `Cmd.output` in #463, and actually ported these bits in this PR, but I don't recall seeing code that calls `opam env`.